### ISSUE TITLE
After converting the bert-base-Chinese model to fp16, the accuracy drops greatly.

### ISSUE THREAD

**Issue by shhn1** [NVIDIA/TensorRT#2585]
## Environment

**TensorRT Version**: 8.5.1
**NVIDIA GPU**: V100
**NVIDIA Driver Version**: 470.103.01
**CUDA Version**: 11.8

When I converted the bert-base-Chinese onnx model to tensorrt fp32, the accuracy dropped very little, but why the accuracy dropped a lot when it was converted to tensorrt fp16？

**Comment by zerollzeng** [comment]
A similar issue like https://github.com/NVIDIA/TensorRT/issues/1196. Setting some layers back to FP32 should be a feasible solution. cc @ttyio

**Comment by shhn1** [comment]
I have tried the steps in issue #1196. My steps are as follows:
    1.` config.set_flag(trt.BuilderFlag.FP16)`
    2. `config.flags = config.flags | 1<<int(trt.BuilderFlag.STRICT_TYPES)`
    3. ```
for i in range(network.num_layers):
    if network.get_layer(i).type == trt.LayerType.SOFTMAX:
        network.get_layer(i).precision = trt.DataType.FLOAT
        network.get_layer(i).get_output(0).dtype = trt.DataType.FLOAT
```
After taking the above steps, the accuracy of the converted trt fp16 model drops very little, but the speed is the same as trt fp32. I don't know why this happens.

**Comment by zerollzeng** [comment]
Can you provide the full TRT verbose log(full fp16 and mixed precision log)? it should tell you why the perf dropped.

**Comment by shhn1** [comment]
The following two are the full TRT verbose log. Thanks!
[build-engine-fp16.txt](https://github.com/NVIDIA/TensorRT/files/10359334/build-engine-fp16.txt)
[build-engine-mixed-precision.txt](https://github.com/NVIDIA/TensorRT/files/10359336/build-engine-mixed-precision.txt)

**Comment by shhn1** [comment]
In `build-engine-mixed-precision.txt`, I got some warning infos as follows：
……
[01/06/2023-10:10:28] [TRT] [W] onnx2trt_utils.cpp:403: One or more weights outside the range of INT32 was clamped
……
[01/06/2023-10:10:29] [TRT] [W] Tensor DataType is determined at build time for tensors not marked as input or output.
……
[01/06/2023-10:10:29] [TRT] [W] Tensor DataType is determined at build time for tensors not marked as input or output.
……
[01/06/2023-10:11:53] [TRT] [V] --------------- Timing Runner: {ForeignNode[bert_model.bert.embeddings.position_embeddings.weight.../Sub]} (Myelin)
[01/06/2023-10:11:53] [TRT] [W] Skipping tactic 0x0000000000000000 due to Myelin error: [verify_output_type] Mismatched type for tensor /Slice_output_0', f16 vs. expected type:f32.
[01/06/2023-10:11:53] [TRT] [V] Fastest Tactic: 0xd15ea5edd15ea5ed Time: inf
……
[01/06/2023-10:12:03] [TRT] [W] TensorRT encountered issues when converting weights between types and that could affect accuracy.
[01/06/2023-10:12:03] [TRT] [W] If this is not the desired behavior, please modify the weights or retrain with regularization to adjust the magnitude of the weights.
[01/06/2023-10:12:03] [TRT] [W] Check verbose logs for the list of affected weights.
[01/06/2023-10:12:03] [TRT] [W] - 122 weights are affected by this issue: Detected subnormal FP16 values.
[01/06/2023-10:12:03] [TRT] [V]   List of affected weights: /Constant_5_output_0 + (Unnamed Layer* 1897) [Shuffle], /Constant_9_output_0 + (Unnamed Layer* 1913) [Shuffle], bert_model_bert_embeddings_position_embeddings_weight_constant, bert_model_bert_embeddings_token_type_embeddings_weight_constant, bert_model_bert_embeddings_word_embeddings_weight_constant, bert_model_bert_encoder_layer_0_attention_self_key_bias _ (Unnamed Layer_ 75) [Shuffle]_constant, bert_model_bert_encoder_layer_10_attention_output_dense_bias _ (Unnamed Layer_ 1577) [Shuffle]_constant, bert_model_bert_encoder_layer_10_attention_self_key_bias _ (Unnamed Layer_ 1505) [Shuffle]_constant, bert_model_bert_encoder_layer_10_attention_self_value_bias _ (Unnamed Layer_ 1523) [Shuffle]_constant, ……
……
[01/06/2023-10:12:03] [TRT] [W] - 72 weights are affected by this issue: Detected values less than smallest positive FP16 subnormal value and converted them to the FP16 minimum subnormalized value.
[01/06/2023-10:12:03] [TRT] [V]   List of affected weights: /bert_model/bert/embeddings/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 55) [Shuffle], /bert_model/bert/encoder/layer.0/attention/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 157) [Shuffle], /bert_model/bert/encoder/layer.0/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 198) [Shuffle], /bert_model/bert/encoder/layer.1/attention/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 300) [Shuffle], /bert_model/bert/encoder/layer.1/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 341) [Shuffle], /bert_model/bert/encoder/layer.10/attention/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 1587) [Shuffle], /bert_model/bert/encoder/layer.10/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 1628) [Shuffle], /bert_model/bert/encoder/layer.11/attention/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 1730) [Shuffle], /bert_model/bert/encoder/layer.11/output/LayerNorm/Constant_1_output_0 + (Unnamed Layer* 1771) [Shuffle], ……
……
I'm confused about the meaning of the above warning infos. Are they causing the drop in perf?

**Comment by zerollzeng** [comment]
Below is the engine information of you 2 configs
[build-engine-fp16.txt](https://github.com/NVIDIA/TensorRT/files/10359334/build-engine-fp16.txt)
```
[01/06/2023-10:03:55] [V] [TRT] Engine Layer Information:
Layer(ShapeHostToDevice): [HostToDeviceCopy 0], Tactic: 0x0000000000000000,  -> token_type_ids[implicit padding mask tensor 0] (Int32[])
Layer(Myelin): {ForeignNode[bert_model.bert.embeddings.position_embeddings.weight.../Sub]}, Tactic: 0x0000000000000000, token_type_ids (Int32[-1,384]), input_ids (Int32[-1,384]), attention_mask (Int32[-1,384]), token_type_ids[implicit padding mask tensor 0] (Int32[]) -> (Unnamed Layer* 1858) [Softmax]_output (Half[-1,4]), /Sub_output_0 (Half[-1])
Layer(TopK): /ArgMax, Tactic: 0x0000000000000000, (Unnamed Layer* 1858) [Softmax]_output (Half[-1,4]) -> (Unnamed Layer* 1861) [TopK]_output_1 (Half[-1,1]), (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1])
Layer(NoOp): (Unnamed Layer* 1865) [Shuffle], Tactic: 0x0000000000000000, (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1]) -> /ArgMax_output_0 (Int32[-1])
Layer(Cast): /Cast, Tactic: 0x0000000000000000, /ArgMax_output_0 (Int32[-1]) -> /Cast_output_0 (Float[-1])
Layer(Reformat): Reformatting CopyNode for Input Tensor 1 to PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2), Tactic: 0x0000000000000000, /Sub_output_0 (Half[-1]) -> Reformatted Input Tensor 1 to PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2) (Float[-1])
Layer(PointWiseV2): PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2), Tactic: 0x0000000000000009, /Cast_output_0 (Float[-1]), Reformatted Input Tensor 1 to PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2) (Float[-1]) -> /Add_2_output_0 (Float[-1])
Layer(NoOp): /Reshape, Tactic: 0x0000000000000000, /Add_2_output_0 (Float[-1]) -> logits (Float[-1,1])
```
[build-engine-mixed-precision.txt](https://github.com/NVIDIA/TensorRT/files/10359336/build-engine-mixed-precision.txt) the log is not complete.
```
[01/06/2023-10:12:03] [TRT] [V] Engine Layer Information:
Layer(ShapeHostToDevice): [HostToDeviceCopy 0], Tactic: 0x0000000000000000,  -> token_type_ids[implicit padding mask tensor 0] (Int32[])
Layer(Myelin): {ForeignNode[bert_model.bert.embeddings.position_embeddings.weight.../Sub]}, Tactic: 0x0000000000000000, token_type_ids (Int32[-1,384]), input_ids (Int32[-1,384]), attention_mask (Int32[-1,384]), token_type_ids[implicit padding mask tensor 0] (Int32[]) -> (Unnamed Layer* 1858) [Softmax]_output (Float[-1,4]), /Sub_output_0 (Float[-1])
Layer(TopK): /ArgMax, Tactic: 0x0000000000000003, (Unnamed Layer* 1858) [Softmax]_output (Float[-1,4]) -> (Unnamed Layer* 1861) [TopK]_output_1 (Float[-1,1]), (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1])
Layer(NoOp): (Unnamed Layer* 1865) [Shuffle], Tactic: 0x0000000000000000, (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1]) -> /ArgMax_output_0 (Int32[-1])
Layer(Cast): /Cast, Tactic: 0x0000000000000000, /ArgMax_output_0 (Int32[-1]) -> /Cast_output_0 (Float[-1])
Layer(PointWiseV2): PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2), Tactic: 0x0000000000000005, /Cast_output_0 (Float[-1]), /Sub_output_0 (Float[-1]) -> /Add_2_output_0 (Float[-1])
Layer(NoOp): /Reshape, Tactic: 0x0000000000000000, /Add_2_output_0 (Float[-1]) -> logits (Float[-1,1])
```
I think the reason why you see the mixed presion engine has the same performance as FP32 is when you fallback the Softmax to FP32, it cause the full myelin subgraph fallback to FP32(`[01/06/2023-10:12:03] [TRT] [V] Engine Layer Information:
Layer(ShapeHostToDevice): [HostToDeviceCopy 0], Tactic: 0x0000000000000000,  -> token_type_ids[implicit padding mask tensor 0] (Int32[])
Layer(Myelin): {ForeignNode[bert_model.bert.embeddings.position_embeddings.weight.../Sub]}, Tactic: 0x0000000000000000, token_type_ids (Int32[-1,384]), input_ids (Int32[-1,384]), attention_mask (Int32[-1,384]), token_type_ids[implicit padding mask tensor 0] (Int32[]) -> (Unnamed Layer* 1858) [Softmax]_output (Float[-1,4]), /Sub_output_0 (Float[-1])
Layer(TopK): /ArgMax, Tactic: 0x0000000000000003, (Unnamed Layer* 1858) [Softmax]_output (Float[-1,4]) -> (Unnamed Layer* 1861) [TopK]_output_1 (Float[-1,1]), (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1])
Layer(NoOp): (Unnamed Layer* 1865) [Shuffle], Tactic: 0x0000000000000000, (Unnamed Layer* 1861) [TopK]_output_2 (Int32[-1,1]) -> /ArgMax_output_0 (Int32[-1])
Layer(Cast): /Cast, Tactic: 0x0000000000000000, /ArgMax_output_0 (Int32[-1]) -> /Cast_output_0 (Float[-1])
Layer(PointWiseV2): PWN(PWN(/Constant_output_0 + (Unnamed Layer* 1868) [Shuffle], /Div), /Add_2), Tactic: 0x0000000000000005, /Cast_output_0 (Float[-1]), /Sub_output_0 (Float[-1]) -> /Add_2_output_0 (Float[-1])
Layer(NoOp): /Reshape, Tactic: 0x0000000000000000, /Add_2_output_0 (Float[-1]) -> logits (Float[-1,1])`)

**Comment by zerollzeng** [comment]
@nvpohanh Is is a known issue for myelin?

**Comment by zerollzeng** [comment]
> I'm confused about the meaning of the above warning infos. Are they causing the drop in perf?

No it won't affect the performance but it may affect the accuracy.

**Comment by nvpohanh** [comment]
> Myelin error: [verify_output_type] Mismatched type for tensor /Slice_output_0', f16 vs. expected type:f32.

Yes, I have seen similar issues before. Let's take this internal for discussion. @zerollzeng FYI, bug 3921137 and 3927970

**Comment by shhn1** [comment]
**[01/06/2023-10:12:03] [TRT] [W] - 122 weights are affected by this issue: Detected subnormal FP16 values.
[01/06/2023-10:12:03] [TRT] [V] List of affected weights: /Constant_5_output_0 + (Unnamed Layer* 1897) [Shuffle], /Constant_9_output_0 + (Unnamed Layer* 1913) [Shuffle], bert_model_bert_embeddings_position_embeddings_weight_constant, bert_model_bert_embeddings_token_type_embeddings_weight_constant, bert_model_bert_embeddings_word_embeddings_weight_constant, bert_model_bert_encoder_layer_0_attention_self_key_bias _ (Unnamed Layer_ 75) [Shuffle]constant, bert_model_bert_encoder_layer_10_attention_output_dense_bias _ (Unnamed Layer 1577) [Shuffle]constant, bert_model_bert_encoder_layer_10_attention_self_key_bias _ (Unnamed Layer 1505) [Shuffle]constant, bert_model_bert_encoder_layer_10_attention_self_value_bias _ (Unnamed Layer 1523) [Shuffle]_constant, ……**

I only fallback the layers mentioned in the above warning infos to fp32, but after building the engine, these warning infos are still displayed. And the model has not changed. I don't kown why this happens?

**Comment by zerollzeng** [comment]
Did you set the layer precision constraint flag, e.g. `--precisionConstraints=obey` in trtexec. You can refer to https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#trtexec-flags and the API docs.

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, thank you!
